generator/src/main/java/org/apache/kafka/message/FieldSpec.java

```diff
@@ -157,11 +157,6 @@ private void checkTagInvariants() {
                     this.taggedVersions + ", which is not open-ended.  taggedVersions must " +
                     "be either none, or an open-ended range (that ends with a plus sign).");
             }
-            if (!this.taggedVersions.intersect(this.versions).equals(this.taggedVersions)) {
-                throw new RuntimeException("Field " + name + " specifies taggedVersions " +
-                    this.taggedVersions + ", and versions " + this.versions + ".  " +
-                    "taggedVersions must be a subset of versions.");
-            }
```

Comment:
I see that without this we get this error when trying to process the json:
```
Exception in thread "main" java.lang.RuntimeException: Exception while processing src/main/resources/common/metadata/PartitionChangeRecord.json
        at org.apache.kafka.message.MessageGenerator.processDirectories(MessageGenerator.java:248)
        at org.apache.kafka.message.MessageGenerator.main(MessageGenerator.java:367)
Caused by: com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `org.apache.kafka.message.FieldSpec`, problem: Field Replicas specifies taggedVersions 0+, and versions 0.  taggedVersions must be a subset of versions.
 at [Source: (File); line: 36, column: 82] (through reference chain: org.apache.kafka.message.MessageSpec["fields"]->java.util.ArrayList[4])
```

However, if we try to specify it with `"taggedVersions": "0"`, like this:
```
    { "name": "Replicas", "type": "[]int32", "default": "null", "entityType": "brokerId",
      "versions": "0", "nullableVersions": "0+", "taggedVersions": "0", "tag": 2,
      "about": "null if the replicas didn't change; the new replicas otherwise." },

```

We get a different error:
```
Exception in thread "main" java.lang.RuntimeException: Exception while processing src/main/resources/common/metadata/PartitionChangeRecord.json
        at org.apache.kafka.message.MessageGenerator.processDirectories(MessageGenerator.java:248)
        at org.apache.kafka.message.MessageGenerator.main(MessageGenerator.java:367)
Caused by: com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `org.apache.kafka.message.FieldSpec`, problem: Field Replicas specifies taggedVersions 0, which is not open-ended.  taggedVersions must be either none, or an open-ended range (that ends with a plus sign).
 at [Source: (File); line: 36, column: 82] (through reference chain: org.apache.kafka.message.MessageSpec["fields"]->java.util.ArrayList[4])
```

The processing code is assuming that a tagged field can never be removed.  I do not understand why this constraint holds.  @cmccabe could you comment?

Reply:
Tagged fields cannot be removed because reusing a tag would cause incompatibility. Eventually we should have a way of "burning" a tag, but we don't have it yet. So do not remove tagged fields for now.

Reply:
We are not removing the tagged field. `"tag": 2` stays reserved to the `Replicas` field, which prevents it from being reused. We're only closing the `versions` range.

https://github.com/apache/kafka/pull/14290/files#diff-b157273749dc11f19aa59bd9ba179f9d9d36f22a30b2ec9aef6491aadb3f9942L33-R36

Closing the version range has desirable consequences:

* removes the field from the generated schema
* prevents serializing the field outside the version range
* generates a runtime error if deserializing the field outside the version range
* marks deprecation of the field in the JSON definition

Even if a tagged field is no longer used we still need to keep the tag reserved to prevent reuse, so the rule "taggedVersions must be a subset of versions." enforced by the removed validation above doesn't seem to make sense? Am I missing some downside to removing it?

Reply:
Yeah, I think that's correct, the .json file still contains the now-closed tagged field and its tag.  I wonder if that is good enough to "burn" that tag.  Seems like it should be, but I am not familiar enough with these details to know.  @cmccabe any thoughts?

Reply:
Yes, this is a good point. I am OK with changing the code so that tagged fields can have an "end version". It would be good to know that all v1 PartitionChangeRecords must specify a storage directory array if they specify a replica array.

However, that isn't quite what the code above is enforcing ... the code above is enforcing that "taggedVersions must be a subset of versions" which does seem like an invariant we still want. I think the code that throws an exception if taggedVersions has an end version is somewhere else...

Reply:
> It would be good to know that all v1 PartitionChangeRecords must specify a storage directory array if they specify a replica array.

If the new field is specified — named `Assignment` (`[]ReplicaAssignment`), replacing `Replicas` (`[]int32`) — it must include `Broker` (`int32`) and `Directory` (`uuid`) which is not a tagged field, so it must be specified.

> I think the code that throws an exception if taggedVersions has an end version is somewhere else...

It makes sense to keep that, and the check you are referring to is still there: https://github.com/apache/kafka/blob/trunk/generator/src/main/java/org/apache/kafka/message/FieldSpec.java#L155-L159

> the code above is enforcing that "taggedVersions must be a subset of versions" which does seem like an invariant we still want.

"taggedVersions must be a subset of versions"  is essentially two checks:

1. "lowest _taggedVersion_  >= lowest _version_"; and
2. "highest _taggedVersion_ is <= highest _version_"

Check 2. is at odds with the requirement that _taggedVersions_ must not have an end version. But we can keep the first check. Instead of removing the code above, we can change it so that we verify that the lowest _tagged version_ is equal to or lower than the _version_ e.g.

```java
if (this.taggedVersions.lowest() > this.versions.highest()) {
    throw new RuntimeException("Field " + name + " specifies taggedVersions " +
            this.taggedVersions + ", and versions " + this.versions + ".  " +
            "The lowest taggedVersion must be lower or equal to the highest version.");
}
```

WDYT?

Reply:
I've made the change above. PTAL

Reply:
It's best to figure out how to burn tagged fields in a separate change. I've reverted the changes to the generator and to the deprecation of the Replicas field so we can keep both of them in the new record versions for now.